### DISCUSSION TITLE
Unknown generic parameter and TypeVar `bound` field

### DISCUSSION THREAD

**Discussion by math-fehr** [microsoft/pyright#3891] [Q&A] [answered]
Hello,

When I check if a variable is of a certain generic type, the generic type parameters appear as unknown, even when they are bound in the respective TypeVar. Here is a minimal example of what I mean:
```python3
from typing import Any, TypeVar, Generic
from dataclasses import dataclass

A = TypeVar("A", bound = int | str)

@dataclass
class V(Generic[A]):
    a: A

def foo(v: Any):
    if isinstance(v, V):      # v is of type V[Unknown] here
        print(v.a)            # This triggers reportUnknownMemberType
                              # I would have (probably wrongly) expected `v.a`
                              # to be of type `int | str`
```
How could I get Pyright to know that that `v.a` is of type `int | str`? Essentially, that the generic parameter is correctly bound to the right type.

**Comment by erictraut** [comment, marked as answer]
The `bound` is an upper bound type for the type variable. If no `bound` is specified for a TypeVar, then the upper bound is implicitly `object`. The type captured by a type variable can be any subtype of the bound. It wouldn't be safe for the type checker to assume that the type variable should be _equal_ to the bound type. That's why it's necessary to assume `Unknown` in this case. 

Consider this example:
```python
class Parent: ...
class Child(Parent): ...

A = TypeVar("A", bound=Parent)

class V(Generic[A]):
    a: A

def set_parent(v: V[Parent]):
    v.a = Parent()

def foo(v: Any):
    if isinstance(v, V):
        set_parent(v)

v1 = V[Child]()
foo(v1)
assert isinstance(v1.a, Child) # Assertion fails!
```

If the `isinstance` type guard narrowed `v` to `V[Parent]`, the above would type check without any errors, even in strict mode. But you can see that there's a type violation here. The type of `v1` is `V[Child]`, but it is incorrectly modified such that `v1.a` contains a `Parent` instance after the call to `foo` returns.

Now, to answer your question about a workaround...

First, I'd ask whether it's really necessary for `foo` to accept an argument of any type. Could you constrain it to a union of specific types (e.g. `V[Any] | W[Any]`)? If so, the `isinstance` type narrowing logic will handle this properly.

If that's not possible, you might need to resort to a cast.
```python
        v1 = cast(V[int | str], v)
        print(v1.a)
```

Or you could use a `# pyright: ignore` to suppress the error:
```python
    if isinstance(v, V):
        print(
            v.a  # pyright: ignore[reportUnknownMemberType, reportUnknownArgumentType]
        )
```

**Reply by math-fehr** [reply]
Thanks a lot! This makes perfect sense now. I just have another question then.

I'm in the case where `V` is a frozen dataclass. Do you know if in that case a similar problem might also happen, since in your counterexample, the problem comes from mutability of one of the fields.
If this is not the case, would it make sense to add a feature request to pyright so that pyright would infer `V[Parent]` here, or would that be too specific to be added?

**Reply by erictraut** [reply]
I'm not sure I understand what you're proposing. Could you provide a self-contained code sample?

**Reply by math-fehr** [reply]
Yes, here is an example of what I mean:
```
class Parent: ...
class Child(Parent): ...

A = TypeVar("A", bound=Parent)

@dataclass(frozen=True)               # V cannot be modified after creation
class V(Generic[A]):
    a: A

def foo(v: Any):
    if isinstance(v, V):                        # Here, is it safe to infer the type `V[Parent]` to `v` ?
        print(v.a)
```
From my understanding, since we cannot reassign `a`, infering the type `V[Parent]` for `v` should be correct, right?

What I'm wondering (if my assumption is true), is if this would make sense to be added to pyright, since this seems quite specific.

**Reply by erictraut** [reply]
Ah, I see what you mean.  This isn't "type inference" but rather "type narrowing".

I'm still not sure that it would be type safe to narrow the type to `V[Parent]` if the type parameter were covariant rather than invariant (as it is currently). I'd need to think about that more. But what you're suggesting would be a really significant change to multiple complex parts of pyright to handle a really special case, so it's not something I'd be inclined to do. Sorry, but you'll need to use a workaround.

**Reply by math-fehr** [reply]
Yes, I understand!
Thanks for explaining me the problem and showing me the workarounds anyway!
